mlflow/utils/autologging_utils.py

```diff
@@ -113,3 +118,65 @@ def wrap_patch(destination, name, patch, settings=None):
     wrapped = functools.wraps(original)(patch)
     patch = gorilla.Patch(destination, name, wrapped, settings=settings)
     gorilla.apply(patch)
+
+
+class _InputExampleInfo:
+    def __init__(self, input_example=None, error_msg=None):
+        self.input_example = input_example
+        self.error_msg = error_msg
+
+
+def handle_input_example_and_signature(
+    get_input_example, infer_model_signature, log_input_example, log_model_signature, logger
+):
+    """
+    Handles the logic of calling functions to gather the input example and infer the model
+    signature.
+
+    :param get_input_example: function which returns an input example, usually sliced from a
+                              dataset. This function can raise an exception, its message will be
+                              shown to the user in a warning in the logs.
+    :param infer_model_signature: function which takes an input example and returns the signature
+                                  of the inputs and outputs of the model. This function can raise
+                                  an exception, its message will be shown to the user in a warning
+                                  in the logs.
+    :param log_input_example: whether to log errors while collecting the input example, and if it
+                              succeeds, whether to return the input example to the user. We collect
+                              it even if this parameter is False because it is needed for inferring
+                              the model signature.
+    :param log_model_signature: whether to infer and return the model signature.
+    :param logger: the logger instance used to log warnings to the user during input example
+                   collection and model signature inference.
+
+    :return: A tuple of input_example and signature. Either or both could be None based on the
+             values of log_input_example and log_model_signature.
+    """
```

Comment:
I know we discussed this offline, but can we insert some logic to short-circuit & return None, None if both of these conditions are false? Just in case sampling that input data is expensive for some reason.

Reply:
we already have this: the get_input_example only fires `if log_input_example or log_model_signature:`

Reply:
Doh!